language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -228,43 +231,54 @@ typeOf (TyInst x t ty) = do
             k' <- kindOf ty
             typeCheckStep
             if k == k'
-                then pure (tyReduce (tySubstitute (extractUnique n) (void ty) ty''))
+                then tyReduce =<< tySubstitute (extractUnique n) (void ty) ty''
                 else throwError (KindMismatch x (void ty) k k')
         _ -> throwError (TypeMismatch x (void t) (TyForall () dummyTyName dummyKind dummyType) (void ty'))
 typeOf (Unwrap x t) = do
     ty <- typeOf t
     case ty of
         TyFix _ n ty' -> do
-            let subst = tySubstitute (extractUnique n) ty ty'
-            pure (tyReduce subst)
+            subst <- tySubstitute (extractUnique n) ty ty'
+            typeCheckStep
```

Comment:
Somewhat unrelated, but I'm confused about what the policy is about where these go.

Reply:
I believe @effectfully suggested that I  use `Language.PlutusCore.Constant.Typed`.

Reply:
I mean the `typeCheckStep`s? Are we talking about the same thing?

Reply:
Oh, no I was confused. I believe they accompany a `tySubstitute`. Perhaps I should put them before?

Reply:
I don't know - hence why I asked what the policy was! I think we could do with a note somewhere in this module about what things we consider to consume gas - but that can be done separately.